database-plugins/src/main/java/io/cdap/plugin/DBRecord.java

```diff
@@ -124,23 +137,37 @@ public void readFields(ResultSet resultSet) throws SQLException {
       int sqlType = metadata.getColumnType(i + 1);
       int sqlPrecision = metadata.getPrecision(i + 1);
       int sqlScale = metadata.getScale(i + 1);
+      // handles backwards compatibility
+      boolean handleAsDecimal = false;
+      Schema outputFieldSchema = field.getSchema();
+      outputFieldSchema = outputFieldSchema.isNullable() ? outputFieldSchema.getNonNullable() : outputFieldSchema;
+      if (sqlType == Types.DECIMAL && (outputSchema == null ||
+        outputFieldSchema.getLogicalType() == Schema.LogicalType.DECIMAL)) {
+        handleAsDecimal = true;
+      }
       setField(resultSet, recordBuilder, field, sqlType, sqlPrecision, sqlScale, nameMap.getOrDefault(field.getName(),
-                                                                                                      field.getName()));
+                                                                                                      field.getName()),
+               handleAsDecimal);
     }
     record = recordBuilder.build();
   }
 
   private void setField(ResultSet resultSet, StructuredRecord.Builder recordBuilder, Schema.Field field, int sqlType,
-                        int sqlPrecision, int sqlScale, String originalName) throws SQLException {
+                        int sqlPrecision, int sqlScale, String originalName, boolean handleAsDecimal)
+    throws SQLException {
     // original name has to be used to get result from result set
-    Object o = DBUtils.transformValue(sqlType, sqlPrecision, sqlScale, resultSet, originalName);
+    Object o = DBUtils.transformValue(sqlType, sqlPrecision, sqlScale, resultSet, originalName,
+                                      handleAsDecimal);
```

Comment:
There is no need for `handleAsDecimal ` variable in this method. We can just pass `Schema.Field field` to the method and get the type of field to derive if decimal type needs to be handled.